internal/storage/blob/store.go

```diff
@@ -214,70 +232,67 @@ func (s *Store) init(ctx context.Context) error {
 	return nil
 }
 
-func (s *Store) clone(ctx context.Context) (*CloneResult, error) {
-	ctx, cancelFunc := s.conf.getCloneCtx(ctx)
-	defer cancelFunc()
-
-	return s.cloner.Clone(ctx)
-}
-
 func (s *Store) updateIndex(ctx context.Context) error {
 	s.log.Debug("Checking for updates")
-
-	changes, err := s.clone(ctx)
+	cr, err := s.clone(ctx)
 	if err != nil {
-		return err
-	}
-
-	if failures := changes.failures(); failures > 0 {
-		s.log.Warnf("Failed to download (%d) files", failures)
+		return fmt.Errorf("failed to clone blob store: %w", err)
 	}
 
-	if changes.isEmpty() {
+	if cr.isEmpty() {
 		s.log.Debug("No changes")
 		return nil
 	}
 
-	s.log.Infof("Detected changes: added or updated (%d), deleted (%d)", len(changes.updateOrAdd), len(changes.delete))
+	s.log.Infof("Detected changes: added or updated (%d), deleted (%d)", len(cr.addedOrUpdated), len(cr.deleted))
+	dotCacheDir := filepath.Join(s.conf.WorkDir, dotcache)
+	workDir := filepath.Join(s.conf.WorkDir, strconv.FormatInt(time.Now().Unix(), 10))
+	if err := validateOrCreateDir(workDir); err != nil {
+		return fmt.Errorf("failed to create new work directory %s: %w", workDir, err)
+	}
 
-	var p *policyv1.Policy
-	var event storage.Event
-	for _, f := range changes.updateOrAdd {
-		if schemaFile, ok := util.RelativeSchemaPath(f.file); ok {
-			s.NotifySubscribers(storage.NewSchemaEvent(storage.EventAddOrUpdateSchema, schemaFile))
-			continue
+	for file, etag := range cr.all {
+		source := filepath.Join(workDir, string(file))
+		dest := filepath.Join(dotCacheDir, string(etag))
+		if err := s.createSymlink(dest, source); err != nil {
+			return fmt.Errorf("failed to create symbolic links for the cloned file: %w", err)
 		}
+	}
 
-		p, err = policy.ReadPolicyFromFile(s.fsys, f.file)
-		if err != nil {
-			return err
-		}
-		entry := index.Entry{File: f.file, Policy: policy.Wrap(policy.WithSourceAttributes(p, driverSourceAttr, etagSourceAttr(f.etag)))}
-		event, err = s.idx.AddOrUpdate(entry)
-		if err != nil {
-			return err
+	idx, err := index.Build(ctx, os.DirFS(workDir), index.WithRootDir("."), index.WithSourceAttributes(driverSourceAttr))
+	if err != nil {
+		if rerr := os.RemoveAll(workDir); rerr != nil && !os.IsNotExist(rerr) {
+			return fmt.Errorf("failed to remove the directory %s: %w", s.workDir, errors.Join(rerr, err))
```

Comment:
This part is nearly identical to initial clone above and is kind of critical for correct functioning. I think it'd be better to define it as a reusable function if possible.